### PR TITLE
Re-add support for macOS

### PR DESCRIPTION
**Pull request by philippremy**
## Project support for macOS

This is now ready to be merged ^^.

I know that these look like some heavy changes, but I believe that I did not break any Windows/Linux infrastructure. I tried my best to do CI on both Windows and Linux, and at least with the official GitHub CI, they appeared to work. Compiling embedded dependencies on Linux works as expected, if you don't attempt to build an embedded Assimp - but that issue was already present before (on recent GCCs, one header is missing a `#include <cstdint>`, which is a one-line fix).

### Overview of changes

Each commit has its own description on what changed and why that change is required in my opinion. But as a quick overview:

 - **OFA**: The whole OptimizeForArchitecture infrastructure has been reworked. Previously, only `x86_64` was properly supported, now there is very solid `arm/arm64`, expanded `x86_64` and rudimentary `ppc` support.
 - **Dependencies**: Various embedded dependency version liftings - all changes here appear ABI and API compatible, but mainly solve build issues on macOS present in the older versions (see the individual commit messages). This does *not* raise the minimum versions required globally, so the current minimum dependency rules still apply.
 - **macOS bundler**: I wrote a small Python tool for creating a standalone bundle on macOS. This allows to create fully redistributable packages that do not depend on libraries other than those provided by macOS itself.
 - **Configuration variables**: When building embedded dependencies, only a small subset of CMake variables was passed through to the actual AliceVision subproject (`/src`). Now the user can overwrite all variables properly by passing them on the CMake CLI (backwards-comparible).
 - **Remove BLAS/LAPACK and Suitesparse on macOS**: Apple provides `Accelerate.framework`, which Ceres can use instead of Suitesparse. This lifts the requirement for a Fortran compiler on macOS and drastically reduces compilation times and bundle size. When running the unit tests I did not notice any performance differences.
 - **CUDA**: As CUDA is completely unsupported on Apple platforms, all related functionality is properly guarded behind conditional checks in the build infrastructure.
- **Cross-compiling**: As cross-compiling on macOS is usually a matter of passing `-arch <ARCH>` to the compiler, the project fully supports cross-compiling from `arm64` to `x86_64` on macOS - including all required dependencies. The target architecture can be set by specifying `CMAKE_OSX_ARCHITECTURES=<arm64|x86_64>` on the CMake CLI. Note that compiling universal binaries is *not* supported at this point, because some dependencies do not have the required logic to do this in one step.
- **Build manual**: A special `INSTALL_macOS.md` provides specific instructions on how to compile this project successfully on macOS.
- **OpenMP**: Using OpenMP with `AppleClang` is now supported - this requires that OpenMP is installed as an external dependency (enabled by default on macOS), as the Apple SDK does not ship it. Furthermore, the project now uses the unified CMake targets `OpenMP::OpenMP_C` and `OpenMP::OpenMP_CXX` respectively instead of setting the required flags manually.

### Open questions

- [x] Reliable Boost build with cross-compiling capabilities
- [x] Remove PCL dependency because it is unused (see #2012). I just opted to not care :D.

### Drawbacks

Because AliceVision has so many dependencies - some with very specific configurations - I think we cannot officially support building the project with package-manager provided dependencies. I just stumbled from roadblock to roadblock when attempting to standardize the required dependencies for the different package managers. So currently the *only* option is to build the project with `ALICEVISION_BUILD_DEPENDENCIES=ON`. However, this matches the recommendation for Linux, so I think this should be fine. If someone wants to attempt to build it with Homebrew/MacPorts/Nix anyway, that might or might not work. But at least we don't have the obligation to ensure compatibility for these cases.

There is no CI yet. I decided against CI, because it currently requires to rebuild all dependencies over and over again - the GitHub runners usually only have 3 threads available, so this is quite slow. I tested pre-built dependencies, but that currently does not work because there are some dependencies hard-code absolute paths into their CMake config modules - making them not portable.

There are no unit tests yet. I ran the unit tests on my Mac, and two are currently failing:
1. `acRansac_test.cpp`: I think there is some undefined behavior here, which caused a segfault on macOS. See [this gist](https://gist.github.com/philippremy/180bcc7d042755206f4261b57af41e11) for a detailed description and proposed solution.
2. `fundamental10PSolver_test.cpp`: The check on line 184 fails. I have no idea why and I don't know anywhere near enough to solve this. Help is much appreciated. For the results see [this gist](https://gist.github.com/philippremy/1c7c1bab3cf9440709c95c60cfa1a542).

### Future plans

This is the first step to make macOS a fully supported platform for this project again. The next steps would be to add the Apple Metal DepthMap backend and maybe generally abstract the DepthMap library to allow for multiple backends with a unified interface.

Also: Add a proper CI solution and ensure that all unit tests pass.

### Feedback is appreciated!

These patches are the result of weeks of debugging, investigating and rebasing - but I am *sure* that there are still issues and edge cases that I completely missed. So fresh viewpoints and remarks are much appreciated. If anyone has the chance to test this out, please do so and let me know if there are any issues.

### CODE REVIEW

**Comment by servantftransperfect** [comment]
Sorry for the long time to answer, i'm still figuring on how to get access to a mac.

**Comment by philippremy** [comment]
No worries, I'll be busy with exams until the end of February anyways, so I won't be able to work on this full time until then. Just wanted to get this out, so at least the RFC period can start :).

**Comment by philippremy** [comment]
A quick follow up on the Boost/LLVM issue:

This appears to be a known upstream problem, which (as of the time of writing) can only be fixed by *overwriting LLVM headers*, see [here](https://github.com/Homebrew/homebrew-core/issues/235411#issuecomment-3732301664) and [here](https://github.com/llvm/llvm-project/issues/77653#issuecomment-3732272052).

So there are essentially three options to approach this (imho):

1. Patch the header during the configure step (horrible idea, not reproducible, tampering with foreign system-wide headers should never be an option)
2. Explicitly restrict usage of package manager LLVM/Clang on macOS to known working versions (not good either, because I don't want to force users to uninstall LLVM just to build the project)
3. Wait until we can drop `b2` as the build system for Boost (I like this one :D, `b2` really is annoying at times):

    The blocking factor currently is CCTag, which relies on the legacy `Boost::math_c99` module to be present, which the CMake build system of Boost cannot build. So we need to wait for either option:

    3.1. Wait until https://github.com/alicevision/CCTag/pull/232 is merged, which removes the module as a dependency.
    3.2. Wait until the Boost.Math maintainer adds the CMake infrastructure for the legacy modules. I've opened an issue and they are willing to add this, although there is no specific timeline yet.

I strongly prefer option 3.1. (as it is the most controllable from the organization's pov), but if you have other ideas: Please let me know ^^.

**Comment by NeverGET** [comment]
Hi @music-dsp-collection — thank you so much for the incredible work on MTL-AliceVision! 🎉

I've been testing the Meshroom 3D viewer with your macOS Metal bundle on Apple Silicon (M4 Max, macOS 26.3) and it's working great. We ran into some Meshroom-side issues where the 3D viewer crashed due to Metal RHI's strict vertex descriptor validation — custom Qt3D geometries (Grid3D, BoundingBox, Locator3D) and loaded meshes didn't always provide the `vertexNormal` attribute that built-in materials expect.

We've put together fixes in a Meshroom PR: alicevision/Meshroom#3030

**Key fixes on the Meshroom side:**
- Remove `vertexNormal` dependency from SphericalHarmonics shaders (compute flat normals via derivatives instead)
- Add default normal attributes to custom geometries
- Add `ensureNormals()` for dynamically loaded meshes
- OpenGL 3.2 Core fallback shader techniques
- macOS platform support in `setupEnvironment()` (DYLD paths, Metal RHI backend)
- QML plugin path discovery for qtAliceVision

Everything is running stable now — the full Meshroom pipeline (camera init → feature extraction → matching → structure from motion → texturing) works end-to-end, and the 3D viewer renders loaded meshes with textures via EXR support from QtAliceVisionImageIO.

Happy to help with any additional macOS testing! Thanks again for making this possible. 🙏

**Comment by philippremy** [comment]
Happy to hear that :D! I finished my exams and will be able to start working on this again shortly. Hopefully I'll be able to iron out the last blockers and turn this into a proper pull request soon :).

**Comment by philippremy** [comment]
### Update: One step closer

> 3\. 3.2. Wait until the Boost.Math maintainer adds the CMake infrastructure for the legacy modules. I've opened an issue and they are willing to add this, although there is no specific timeline yet.

It seems like this is now the most probable option. A PR in the Boost.Math repo is in the process of being merged, so this could actually be resolved very soon. I'll work on switching to the CMake build system for Boost as soon as the interface is stable (i.e., the PR is merged). Once this is done, I think the infrastructure changes are ready to be merged 🚀.

AFAIK the required changes for Boost should land in v1.91.0, which is scheduled to be released on April 8th!

**Comment by philippremy** [comment]
CCTag merged the required PR, so this technically has no blockers left - except that it needs a huge rebase/rework because of the PR that changed the dependency handling.
I will start working on that asap :D.

**Comment by servantftransperfect** [comment]
> CCTag merged the required PR, so this technically has no blockers left - except that it needs a huge rebase/rework because of the PR that changed the dependency handling. I will start working on that asap :D.

My bad, sorry for that. I forgot you had made changes to those files.

**Comment by philippremy** [comment]
No worries, that was something I wanted to tackle anyways. It is muuuch cleaner now and transferring the changes should be as simple as cherrypicking or copy-pasting. I don't expect that to take long :).

**Comment by philippremy** [comment]
I reworked everything onto the new infrastructure and even found some bugs on my side along the way. I tested compilation on my Mac (clean, only required tools per `INSTALL_macOS.md` installed) for arm64 and x86_86 with Unix Makefiles and with Ninja. If CI passes, I will convert this into a proper PR. Bundling also works for both architectures and the bundle now includes the headers, the `share/aliceVision` and the `share/meshroom` folders.

Thanks for the patience everyone ^^!

EDIT: Woops, didn't mean to rewrite *that* much history. Commits are now back with their owners :D.

File: src/cmake/deps/xercesc.cmake [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+    av_add_cmake_dep(
+        TARGET     ${XERCESC_TARGET}
+        SOURCE_DIR xercesc
+        URL        ${DEP_XERCESC_URL}
+        URL_HASH   ${DEP_XERCESC_HASH}
+    )
+
+    set(XERCESC_CMAKE_FLAGS
+      # TODO
```
Comment (servantftransperfect) [inline src/cmake/deps/xercesc.cmake]: status ?

File: src/cmake/Helpers.cmake [hunk 2]
```diff
@@ -213,9 +237,12 @@ function(alicevision_add_software software_name)
 
     set_target_properties(${software_name}_exe
         PROPERTIES SOVERSION ${ALICEVISION_SOFTWARE_VERSION_MAJOR}
-        VERSION "${ALICEVISION_SOFTWARE_VERSION_MAJOR}.${ALICEVISION_SOFTWARE_VERSION_MINOR}"
+        VERSION "${ALICEVISION_SOFTWARE_VERSION_MAJOR}_${ALICEVISION_SOFTWARE_VERSION_MINOR}"
```
Comment (servantftransperfect) [inline src/cmake/Helpers.cmake]: Why ?

Comment (fabiencastan) [reply src/cmake/Helpers.cmake]: The standard is to use dot:
https://cmake.org/cmake/help/latest/prop_tgt/VERSION.html#prop_tgt:VERSION

File: src/cmake/deps/xercesc.cmake (hunk 1 above)
Comment (philippremy) [reply src/cmake/deps/xercesc.cmake]: My bad, this should contain `-DXercesC_DIR:PATH=${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}/cmake/XercesC`.

Should I add a new commit for this or rebase? You can add it, if you want to, of course.

File: src/cmake/Helpers.cmake (hunk 2 above)
Comment (philippremy) [reply src/cmake/Helpers.cmake]: Yes, I have seen the dot notation in many places actually, but so far only for shared libraries. And for these it is not an issue, because they always have the `.dylib` file extension. For Mach-O executables (on macOS) the system went to interpret the last version number as a file extension, because they usually don't have a file extension at all (see the image as an example). While not the biggest issue, e.g. Xcode refuses to load the executable for debugging, because it only checked the file extension (bad practice, I know).

And I thought it would be fine because we created symlinks anyways and as far as I can tell, Meshroom for example will invoke the non-versioned symlinks.

But feel free to reverse, if it breaks stuff I did not notice. Invoking them via the CLI works fine, as far as I know...

file-extension-bug

**Comment by servantftransperfect** [comment]
I can't compile docker images with the updated boost-1.90. Any reason not to stick with 1.86 ? (Except cmake archive)

File: src/cmake/Helpers.cmake (hunk 2 above)
Comment (servantftransperfect) [reply src/cmake/Helpers.cmake]: can't we just do that for mac os ?

Comment (philippremy) [reply src/cmake/Helpers.cmake]: Sounds good to me!

**Comment by philippremy** [comment]
> I can't compile docker images with the updated boost-1.90. Any reason not to stick with 1.86 ? (Except cmake archive)

No, not really - I guess I was still thinking about the `math_c99` which was only introduced very recently for the CMake build system. But because CCTag removed the dependency we could just revert to 1.86 (it has a CMake archive as well and I would really prefer to switch to CMake :D...).

What exactly is causing the build failure in Docker (just out of personal curiosity)?

**Comment by servantftransperfect** [comment]
> > I can't compile docker images with the updated boost-1.90. Any reason not to stick with 1.86 ? (Except cmake archive)
> 
> No, not really - I guess I was still thinking about the `math_c99` which was only introduced very recently for the CMake build system. But because CCTag removed the dependency we could just revert to 1.86 (it has a CMake archive as well and I would really prefer to switch to CMake :D...).
> 
> What exactly is causing the build failure in Docker (just out of personal curiosity)?

I'm getting a weird compilation error in boost maths numeric_limits.hpp. BOOST_STATIC_something contains constexpr which generate a constexpr constexpr function(...).

**Comment by philippremy** [comment]
> I'm getting a weird compilation error in boost maths numeric_limits.hpp. BOOST_STATIC_something contains constexpr which generate a constexpr constexpr function(...).

Uh oh, that doesn't sound good. I think I had the same issue at some point and just forgot about it because I patched it manually. I'll see if I it already fixed upstream and if not, I will open a bug report. The maintainer over there is super helpful and maybe that'll make us future-proof :D.

EDIT: Apparently has been fixed two weeks ago, but that will probably only land in Boost 1.91, because it didn't make it into master in time: https://github.com/boostorg/math/issues/1383.

So if it wasn't a problem in 1.86, we should definitely revert! Using the develop branch for all Boost stuff doesn't seem a good idea imho :).

**Review by servantftransperfect — APPROVED** [review]
That was a pretty long review to do !

So first of all, **thanks for this large PR**. I am sure many people will be happy to see this available.

I reviewed the changes mostly for non regression. 

As it is focused on cmake changes, I checked that our docker containers are still building correctly. The CI is still working correctly.

I am NOT able to validate the APPLE part as we have no access to hardware. 

I did some updates related to the comment i did, and i added some changes for the dockerFiles.

**Comment by philippremy** [comment]
Thank YOU for this awesome project!

The containers are something that I cannot reliably test, so thanks for that as well. For any future PRs, I will try to keep them way smaller.

Everything seems fine to me - but as I mentioned over in my proposal in the discussion section, the build will probably always be somewhat brittle (at least for the mac users under us). And that isn't even our fault, but CMake just really gets to its limits with so many dependencies. I will try to do regular testing for macOS, if possible. And maybe I'll be able to spin up some CI for macOS as well.
